File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Editor/Tools/UnloadSceneTool.cs
using System;
using UnityEditor;
using UnityEditor.SceneManagement;
using UnityEngine.SceneManagement;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for unloading a Unity scene (without deleting the asset)
    /// </summary>
    public class UnloadSceneTool : McpToolBase
    {
        public UnloadSceneTool()
        {
            Name = "unload_scene";
            Description = "Unloads a scene by path or name (does not delete the scene asset, just closes it from the hierarchy)";
        }

        /// <summary>
        /// Execute the UnloadScene tool with the provided parameters
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject</param>
        public override JObject Execute(JObject parameters)
        {
            string scenePath = parameters["scenePath"]?.ToObject<string>();
            string sceneName = parameters["sceneName"]?.ToObject<string>();
            bool removeScene = parameters["removeScene"]?.ToObject<bool?>() ?? true;

            if (string.IsNullOrEmpty(scenePath) && string.IsNullOrEmpty(sceneName))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Provide either 'scenePath' or 'sceneName'",
                    "validation_error"
                );
            }

            try
            {
                Scene sceneToUnload;

                if (!string.IsNullOrEmpty(scenePath))
                {
                    sceneToUnload = SceneManager.GetSceneByPath(scenePath);
                }
                else
                {
                    sceneToUnload = SceneManager.GetSceneByName(sceneName);
                }

                if (!sceneToUnload.IsValid())
                {
                    string identifier = !string.IsNullOrEmpty(scenePath) ? $"path '{scenePath}'" : $"name '{sceneName}'";
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Scene with {identifier} is not currently loaded",
                        "not_found_error"
                    );
                }

                // Check if this is the only loaded scene
                if (SceneManager.sceneCount <= 1)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "Cannot unload the only loaded scene. Load another scene first or create a new scene",
                        "validation_error"
                    );
                }

                string unloadedSceneName = sceneToUnload.name;
                string unloadedScenePath = sceneToUnload.path;
                bool wasDirty = sceneToUnload.isDirty;

                // If scene has unsaved changes, save it first
                if (wasDirty)
                {
                    bool savePrompt = parameters["saveIfDirty"]?.ToObject<bool?>() ?? true;
                    if (savePrompt && !string.IsNullOrEmpty(unloadedScenePath))
                    {
                        EditorSceneManager.SaveScene(sceneToUnload);
                    }
                }

                // Close/unload the scene
                bool success = EditorSceneManager.CloseScene(sceneToUnload, removeScene);

                if (!success)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Failed to unload scene '{unloadedSceneName}'",
                        "unload_error"
                    );
                }

                McpLogger.LogInfo($"Unloaded scene '{unloadedSceneName}' (path: '{unloadedScenePath}')");

                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = $"Successfully unloaded scene '{unloadedSceneName}'",
                    ["sceneName"] = unloadedSceneName,
                    ["scenePath"] = unloadedScenePath,
                    ["wasDirty"] = wasDirty,
                    ["removed"] = removeScene
                };
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error unloading scene: {ex.Message}",
                    "scene_unload_error"
                );
            }
        }
    }
}

File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Server~/src/tools/unloadSceneTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';

const toolName = 'unload_scene';
const toolDescription = 'Unloads a scene by path or name (does not delete the scene asset, just closes it from the hierarchy)';

const paramsSchema = z.object({
  scenePath: z.string().optional().describe("Full asset path to the scene (e.g., 'Assets/Scenes/MyScene.unity')"),
  sceneName: z.string().optional().describe('Scene name without extension (used if scenePath not provided)'),
  saveIfDirty: z.boolean().optional().describe('If true, saves the scene before unloading if it has unsaved changes. Default: true'),
  removeScene: z.boolean().optional().describe('If true, removes the scene from the hierarchy. If false, keeps it but unloaded. Default: true')
});

export function registerUnloadSceneTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any) {
  if (!params.scenePath && !params.sceneName) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Either 'scenePath' or 'sceneName' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to unload scene'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'Successfully unloaded scene'
    }],
    data: {
      sceneName: response.sceneName,
      scenePath: response.scenePath,
      wasDirty: response.wasDirty,
      removed: response.removed
    }
  };
}
